### ISSUE TITLE
Attempt to free unreferenced scalar fiddling with the symbol table

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9351]
Migrated from [rt.perl.org#54934](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=54934) (status was 'resolved')

Searchable as RT54934$

**Comment by p5pRT** [comment]
### From @ntyni


This is a bug report for perl from Niko Tyni \<ntyni@&#8203;debian\.org>\,
generated with the help of perlbug 1\.36 running under perl 5\.10\.0\.




---
As seen in \<http&#8203;://cpantesters\.perl\.org/show/PerlIO\-via\-dynamic\.html>\,
the test suite of the PerlIO&#8203;::via&#8203;::dynamic CPAN module generates 
'Attempt to free unreferenced scalar' warnings with recent perl versions\,
including 5\.10\.0 but not 5\.8\.8\.

The warnings are still there on blead @&#8203;33937; bisecting shows the warnings
started with

 Change 24966 by nicholas@&#8203;ship\-in\-a\-bottle on 2005/06/23 21&#8203;:30&#8203;:33

 Remove the reference loop between symbol tables and typeglobs\. Typeglobs
 now have a weak reference onto their symbol table\.

This snippet is the smallest I can make it\. The original code deletes
a dynamically created package instead which makes more sense than this\.

\#\!/usr/bin/perl
package test;
sub PUSHED \{
&nbsp;    return \\\*FOO;
\}
1;
package main;
\{
&nbsp;    open my $fh\, '\</';
&nbsp;    binmode\($fh\, '&#8203;:via\(test\)'\);
\}
delete $&#8203;::\{'test&#8203;::'\};
\_\_END\_\_



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
Site configuration information for perl 5\.10\.0&#8203;:

Configured by Debian Project at Thu May  8 11&#8203;:57&#8203;:24 UTC 2008\.

Summary of my perl5 \(revision 5 version 10 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.18\-6\-xen\-amd64\, archname=x86\_64\-linux\-gnu\-thread\-multi
&nbsp;    uname='linux sid 2\.6\.18\-6\-xen\-amd64 \#1 smp thu apr 24 05&#8203;:10&#8203;:26 utc 2008 x86\_64 gnulinux '
&nbsp;    config\_args='\-Dusethreads \-Duselargefiles \-Dccflags=\-DDEBIAN \-Dcccdlflags=\-fPIC \-Darchname=x86\_64\-linux\-gnu \-Dprefix=/usr \-Dprivlib=/usr/share/perl/5\.10 \-Darchlib=/usr/lib/perl/5\.10 \-Dvendorprefix=/usr \-Dvendorlib=/usr/share/perl5 \-Dvendorarch=/usr/lib/perl5 \-Dsiteprefix=/usr/local \-Dsitelib=/usr/local/share/perl/5\.10\.0 \-Dsitearch=/usr/local/lib/perl/5\.10\.0 \-Dman1dir=/usr/share/man/man1 \-Dman3dir=/usr/share/man/man3 \-Dsiteman1dir=/usr/local/man/man1 \-Dsiteman3dir=/usr/local/man/man3 \-Dman1ext=1 \-Dman3ext=3perl \-Dpager=/usr/bin/sensible\-pager \-Uafs \-Ud\_csh \-Ud\_ualarm \-Uusesfio \-Uusenm \-DDEBUGGING=\-g \-Doptimize=\-O2 \-Duseshrplib \-Dlibperl=libperl\.so\.5\.10\.0 \-Dd\_dosuid \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-g'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DDEBIAN \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.2\.3 \(Debian 4\.2\.3\-5\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
&nbsp;    libs=\-lgdbm \-lgdbm\_compat \-ldb \-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    perllibs=\-ldl \-lm \-lpthread \-lc \-lcrypt
&nbsp;    libc=/lib/libc\-2\.7\.so\, so=so\, useshrplib=true\, libperl=libperl\.so\.5\.10\.0
&nbsp;    gnulibc\_version='2\.7'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-O2 \-g \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.10\.0&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/local/lib/perl/5\.10\.0
&nbsp;    /usr/local/share/perl/5\.10\.0
&nbsp;    /usr/lib/perl5
&nbsp;    /usr/share/perl5
&nbsp;    /usr/lib/perl/5\.10
&nbsp;    /usr/share/perl/5\.10
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    \.



---
Environment for perl 5\.10\.0&#8203;:
&nbsp;    HOME=/home/niko
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_CTYPE=fi\_FI\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/niko/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/bin&#8203;:/usr/bin/X11&#8203;:/usr/games&#8203;:/sbin&#8203;:/usr/sbin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/zsh

**Comment by p5pRT** [comment]
### From @rgs

2008/5/27 via RT Niko Tyni \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> As seen in \<http&#8203;://cpantesters\.perl\.org/show/PerlIO\-via\-dynamic\.html>\,
> the test suite of the PerlIO&#8203;::via&#8203;::dynamic CPAN module generates
> 'Attempt to free unreferenced scalar' warnings with recent perl versions\,
> including 5\.10\.0 but not 5\.8\.8\.
>
> The warnings are still there on blead @&#8203;33937; bisecting shows the warnings
> started with
>
>  Change 24966 by nicholas@&#8203;ship\-in\-a\-bottle on 2005/06/23 21&#8203;:30&#8203;:33
>
>  Remove the reference loop between symbol tables and typeglobs\. Typeglobs
>  now have a weak reference onto their symbol table\.
>
> This snippet is the smallest I can make it\. The original code deletes
> a dynamically created package instead which makes more sense than this\.
>
> \#\!/usr/bin/perl
> package test;
> sub PUSHED \{
>    return \\\*FOO;
> \}
> 1;
> package main;
> \{
>    open my $fh\, '\</';
>    binmode\($fh\, '&#8203;:via\(test\)'\);
> \}
> delete $&#8203;::\{'test&#8203;::'\};

Good catch\. I've traced the double\-freed scalar; the patch below fixes
the error\. However I'm not sure at all it creates no leak\.

<details><summary>Inline Patch</summary>

```diff
--- ext/PerlIO/via/via.xs.orig	2008-01-25 10:00:53.000000000 +0100
+++ ext/PerlIO/via/via.xs	2008-05-30 14:43:11.000000000 +0200
@@ -89,7 +89,7 @@ PerlIOVia_method(pTHX_ PerlIO * f, const
 	    if (!s->fh) {
 		GV *gv = newGVgen(HvNAME_get(s->stash));
 		GvIOp(gv) = newIO();
-		s->fh = newRV_noinc((SV *) gv);
+		s->fh = newRV((SV *) gv);
 		s->io = GvIOp(gv);
 	    }
 	    IoIFP(s->io) = PerlIONext(f);

```
</details>
End of Patch.

So what leaks is a gv generated to hold the "FOO" reference\. \(Its name
is "\_GEN\_0"\.\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From alexmv@bestpractical.com

On Fri\, 2008\-05\-30 at 14&#8203;:55 \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> Good catch\. I've traced the double\-freed scalar; the patch below fixes
> the error\. However I'm not sure at all it creates no leak\.

This bug also causes segfaults in SVK under 5\.10 and blead\.  Getting
this patch \(or whatever it ends up as\) into 5\.10\.1 would be much
appreciated\.
&nbsp;  Thanks\!
 \- Alex

**Comment by p5pRT** [comment]
### From @ntyni

On Fri\, May 30\, 2008 at 02&#8203;:55&#8203;:22PM \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> 2008/5/27 via RT Niko Tyni \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> > As seen in \<http&#8203;://cpantesters\.perl\.org/show/PerlIO\-via\-dynamic\.html>\,
> > the test suite of the PerlIO&#8203;::via&#8203;::dynamic CPAN module generates
> > 'Attempt to free unreferenced scalar' warnings with recent perl versions\,
> > including 5\.10\.0 but not 5\.8\.8\.

> Good catch\. I've traced the double\-freed scalar; the patch below fixes
> the error\. However I'm not sure at all it creates no leak\.

I'm happy to report that the patch fixes all the numerous failures in
the SVK v2\.0\.2 test suite on Perl 5\.10\.0\.

 http&#8203;://cpantesters\.perl\.org/show/SVK\.html\#SVK\-v2\.0\.2

Many thanks for looking at this\.
\-\- 
Niko Tyni   ntyni@&#8203;debian\.org

**Comment by p5pRT** [comment]
### From @ntyni

On Fri\, May 30\, 2008 at 02&#8203;:55&#8203;:22PM \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> 2008/5/27 via RT Niko Tyni \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> > As seen in \<http&#8203;://cpantesters\.perl\.org/show/PerlIO\-via\-dynamic\.html>\,
> > the test suite of the PerlIO&#8203;::via&#8203;::dynamic CPAN module generates
> > 'Attempt to free unreferenced scalar' warnings with recent perl versions\,
> > including 5\.10\.0 but not 5\.8\.8\.

> Good catch\. I've traced the double\-freed scalar; the patch below fixes
> the error\. However I'm not sure at all it creates no leak\.
> 
> \-\-\- ext/PerlIO/via/via\.xs\.orig    2008\-01\-25 10&#8203;:00&#8203;:53\.000000000 \+0100
> \+\+\+ ext/PerlIO/via/via\.xs    2008\-05\-30 14&#8203;:43&#8203;:11\.000000000 \+0200
> @&#8203;@&#8203; \-89\,7 \+89\,7 @&#8203;@&#8203; PerlIOVia\_method\(pTHX\_ PerlIO \* f\, const
>          if \(\!s\->fh\) \{
>          GV \*gv = newGVgen\(HvNAME\_get\(s\->stash\)\);
>          GvIOp\(gv\) = newIO\(\);
> \-        s\->fh = newRV\_noinc\(\(SV \*\) gv\);
> \+        s\->fh = newRV\(\(SV \*\) gv\);
>          s\->io = GvIOp\(gv\);
>          \}
>          IoIFP\(s\->io\) = PerlIONext\(f\);
> End of Patch\.
> 
> So what leaks is a gv generated to hold the "FOO" reference\. \(Its name
> is "\_GEN\_0"\.\)

Hi\,

I don't see this applied yet\, hope it isn't falling through the cracks\.
While PerlIO\-via\-dynamic\-0\.13 now has a workaround for 5\.10\.0\, Bastian
Blank raised concerns about possible security implications of the double
frees in \<http&#8203;://bugs\.debian\.org/479698> \(Cc'd\)\.

Again\, many thanks for looking at this\.
\-\- 
Niko Tyni   ntyni@&#8203;debian\.org

**Comment by p5pRT** [comment]
### From @rgs

2008/6/6 Niko Tyni \<ntyni@&#8203;debian\.org>&#8203;:
>> \-\-\- ext/PerlIO/via/via\.xs\.orig        2008\-01\-25 10&#8203;:00&#8203;:53\.000000000 \+0100
>> \+\+\+ ext/PerlIO/via/via\.xs     2008\-05\-30 14&#8203;:43&#8203;:11\.000000000 \+0200
>> @&#8203;@&#8203; \-89\,7 \+89\,7 @&#8203;@&#8203; PerlIOVia\_method\(pTHX\_ PerlIO \* f\, const
>>           if \(\!s\->fh\) \{
>>               GV \*gv = newGVgen\(HvNAME\_get\(s\->stash\)\);
>>               GvIOp\(gv\) = newIO\(\);
>> \-             s\->fh = newRV\_noinc\(\(SV \*\) gv\);
>> \+             s\->fh = newRV\(\(SV \*\) gv\);
>>               s\->io = GvIOp\(gv\);
>>           \}
>>           IoIFP\(s\->io\) = PerlIONext\(f\);
>> End of Patch\.
>>
>> So what leaks is a gv generated to hold the "FOO" reference\. \(Its name
>> is "\_GEN\_0"\.\)
>
> Hi\,
>
> I don't see this applied yet\, hope it isn't falling through the cracks\.
> While PerlIO\-via\-dynamic\-0\.13 now has a workaround for 5\.10\.0\, Bastian
> Blank raised concerns about possible security implications of the double
> frees in \<http&#8203;://bugs\.debian\.org/479698> \(Cc'd\)\.

No\. Nicholas commented on IRC that this apparently creates no link\,
and I'm not able to prove him false\, so I'll probably apply this
patch\. \(or someone could beat me to it\)

**Comment by p5pRT** [comment]
### From @rgs

2008/5/30 Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;gmail\.com>&#8203;:
> Good catch\. I've traced the double\-freed scalar; the patch below fixes
> the error\. However I'm not sure at all it creates no leak\.

I've now applied this and bumped the version of PerlIO&#8203;::via\, so this
bug can be closed&#8203;:

Change 34025 by rgs@&#8203;scipion on 2008/06/08 14&#8203;:00&#8203;:59

&nbsp;        Fix for bug \[perl \#54934\] Attempt to free unreferenced scalar
fiddling with the symbol table
&nbsp;        Keep the refcount of the globs generated by PerlIO&#8203;::via balanced\.

Affected files \.\.\.

\.\.\. //depot/perl/ext/PerlIO/via/via\.pm\#9 edit
\.\.\. //depot/perl/ext/PerlIO/via/via\.xs\#17 edit

Differences \.\.\.

==== //depot/perl/ext/PerlIO/via/via\.pm\#9 \(text\) ====

@&#8203;@&#8203; \-1\,5 \+1\,5 @&#8203;@&#8203;
 package PerlIO&#8203;::via;
\-our $VERSION = '0\.05';
\+our $VERSION = '0\.06';
 use XSLoader \(\);
 XSLoader&#8203;::load 'PerlIO&#8203;::via';
 1;

==== //depot/perl/ext/PerlIO/via/via\.xs\#17 \(text\) ====

@&#8203;@&#8203; \-89\,7 \+89\,7 @&#8203;@&#8203;
&nbsp;            if \(\!s\->fh\) \{
&nbsp;                GV \*gv = newGVgen\(HvNAME\_get\(s\->stash\)\);
&nbsp;                GvIOp\(gv\) = newIO\(\);
\-               s\->fh = newRV\_noinc\(\(SV \*\) gv\);
\+               s\->fh = newRV\(\(SV \*\) gv\);
&nbsp;                s\->io = GvIOp\(gv\);
&nbsp;            \}
&nbsp;            IoIFP\(s\->io\) = PerlIONext\(f\);

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'
